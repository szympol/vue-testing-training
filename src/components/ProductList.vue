<script setup lang="ts">
import VButton from "@/components/VButton.vue";
import useFetchProducts from "@/services/fetchProducts.ts";
import HelloWorld from "@/components/HelloWorld.vue";

const { get, loading, results, error } = useFetchProducts();
</script>

<template>
  <div>
    <VButton class="mt-12" @click="get" :loading="loading"
      >Download products</VButton
    >
    <div class="flex justify-center items-center">
      <template v-if="loading">
        <span class="text-blue-800 text-2xl">Loading...</span>
        <svg
          role="status"
          class="w-16 h-16 mr-2 text-gray-200 animate-spin dark:text-gray-600 fill-blue-600 ml-6"
          viewBox="0 0 100 101"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
            fill="currentColor"
          />
          <path
            d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
            fill="currentFill"
          />
        </svg>
      </template>
      <template v-if="error">
        <span class="text-red-800 text-xl">Error...</span>
      </template>
    </div>
    <div class="bg-white">
      <div
        class="max-w-2xl mx-auto py-16 px-4 sm:py-24 sm:px-6 lg:max-w-7xl lg:px-8"
      >
        <h2 class="text-2xl font-extrabold tracking-tight text-gray-900">
          Customers also purchased
        </h2>

        <div
          class="mt-6 grid grid-cols-1 gap-y-10 gap-x-6 sm:grid-cols-2 lg:grid-cols-4 xl:gap-x-8"
        >
          <div
            class="group relative"
            v-for="product in results"
            :key="product.id"
            @click="$emit('selectedProduct', product)"
            data-test-id="product"
          >
            <div
              class="w-full min-h-80 bg-gray-200 aspect-w-1 aspect-h-1 rounded-md overflow-hidden group-hover:opacity-75 lg:h-80 lg:aspect-none"
            >
              <img
                :src="product.image"
                :alt="product.description"
                class="w-full h-full object-center object-cover lg:w-full lg:h-full"
              />
            </div>
            <div class="mt-4 flex justify-between">
              <div>
                <h3 class="text-sm text-gray-700">
                  <a href="#">
                    <span aria-hidden="true" class="absolute inset-0"></span>
                    {{ product.title }}
                  </a>
                </h3>
                <p class="mt-1 text-sm text-gray-500">
                  {{ product.category }}
                </p>
              </div>
              <p class="text-sm font-medium text-gray-900">
                $ {{ product.price }}
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <HelloWorld msg="What a day!" />
  </div>
</template>
