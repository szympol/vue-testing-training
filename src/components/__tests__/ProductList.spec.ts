// import { render, screen, fireEvent } from "@testing-library/vue";
import { describe, it, expect } from "vitest";
import { render, screen, fireEvent } from "@testing-library/vue";
import { mount, flushPromises } from "@vue/test-utils";
import ProductList from "@/components/ProductList.vue";
import { products } from "./mockedData";

// bug with asynchronous queuing
const floodWithFlushPromises = async () => {
  for (let i = 0; i < 50; i++) {
    await flushPromises();
  }
};

describe("product-list", () => {
  it("render products correctly", async () => {
    expect.assertions(8);
    const wrapper = mount(ProductList, {
      global: {
        stubs: {
          HelloWorld: {
            template: "<span class='I am a stub' />",
          },
        },
      },
    });

    // console.log("VUe test utils wrapper: ", wrapper);

    expect(wrapper.text()).not.toContain("Loading");

    const button = wrapper.find("button");
    // const button = wrapper.get('[data-test-id="download-products-button"]');

    await button.trigger("click");

    expect(wrapper.text()).toContain("Loading");

    await floodWithFlushPromises();

    // expect(wrapper.text()).not.toContain("Loading");

    const allProducts = wrapper.findAll('[data-test-id="product"]');

    expect(allProducts).toHaveLength(products.length);

    const firstProduct = allProducts[0];
    const secondProduct = allProducts[1];
    await firstProduct.trigger("click");
    await secondProduct.trigger("click");
    expect(wrapper.emitted()).toHaveProperty("selectedProduct");

    const selectedProductEvent = wrapper.emitted("selectedProduct");
    expect(selectedProductEvent).toHaveLength(2);

    expect((selectedProductEvent as unknown[])[0]).toEqual([products[0]]);
    expect((selectedProductEvent as unknown[])[1]).toEqual([products[1]]);

    const productTitle = products[0].title;

    expect(wrapper.text()).toContain(productTitle);
  });

  it("renders products correctly with testing-library", async () => {
    render(ProductList);

    // const wrapper = render(ProductList);
    // console.log("Testing library wrapper: ", wrapper);

    const productList = await screen.findByText("Download products");

    await fireEvent.click(productList);

    const productTitle = products[0].title;

    const value = await screen.findByText(productTitle);

    expect(value.innerHTML).toContain(productTitle);

    // Training --> please rewrite here the testing implementation from 16-62 lines using testing-library/vue API
  });

  it("matches snapshot", () => {
    const wrapper = mount(ProductList);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
