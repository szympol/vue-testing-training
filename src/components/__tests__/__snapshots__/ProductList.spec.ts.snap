// Vitest Snapshot v1

exports[`product-list > matches snapshot 1`] = `
"<div><button type=\\"button\\" class=\\"text-white bg-blue-700 hover:bg-blue-800 focus:outline-none focus:ring-4 focus:ring-blue-300 font-medium rounded-full text-sm px-5 py-2.5 text-center mr-2 mb-2 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 mt-12\\">Download products</button>
  <div class=\\"flex justify-center items-center\\">
    <!--v-if-->
    <!--v-if-->
  </div>
  <div class=\\"bg-white\\">
    <div class=\\"max-w-2xl mx-auto py-16 px-4 sm:py-24 sm:px-6 lg:max-w-7xl lg:px-8\\">
      <h2 class=\\"text-2xl font-extrabold tracking-tight text-gray-900\\"> Customers also purchased </h2>
      <div class=\\"mt-6 grid grid-cols-1 gap-y-10 gap-x-6 sm:grid-cols-2 lg:grid-cols-4 xl:gap-x-8\\"></div>
    </div>
  </div>
  <div class=\\"greetings\\" data-v-469af010=\\"\\">
    <h1 class=\\"green text-center\\" data-v-469af010=\\"\\">What a day!</h1>
  </div>
</div>"
`;
